packages/btc/tsconfig.json

```diff
@@ -1,20 +1,19 @@
 {
   "compilerOptions": {
-    "moduleResolution": "Node",
+    "moduleResolution": "Bundler",
```

Comment:
The TypeScript document recommends using `"moduleResolution": "Node16"` instead of `"moduleResolution": "Bundler"` for writing a library. 

> In short, "moduleResolution": "bundler" is infectious, allowing code that only works in bundlers to be produced. Likewise, "moduleResolution": "nodenext" is only checking that the output works in Node.js, but in most cases, module code that works in Node.js will work in other runtimes and in bundlers.

See: https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#considerations-for-bundling-libraries

Reply:
> Update "moduleResolution": "NodeNext" in tsconfig.json. This requires all relative imports to contain file extensions, but we would need to refactor all our code to satisfy this requirement.

According to the above comment, the current code has added `.js` extension for the relative imports, so we should use `"moduleResolution": "NodeNext"` instead of `"moduleResolution": "Bundler"`.

Reply:
### My thoughts, at the beginning

According to the TS 5.0+ related documentation, both options are recommended for modern projects. However, after some tests, I think the `Bundler` option is somewhat more reasonable to go with. Though this is not final, so you can continue reading. Any feedback is welcome.

Firstly, I think we have 3 paths to go:
- Specify the packages as `"type": "module"`, and specify `"moduleResolution": "NodeNext"`
- Keep the packages as `"type": "commonjs"`, and specify `"moduleResolution": "NodeNext"`
- Keep the packages as `"type": "commonjs"`, and specify `"moduleResolution": "Bundler"`

And I think the `Bundler` option is more reasonable:
1. We're in fact using `tsup` as the bundler, which transpiles/packs our source code into single CJS/ESM files.
2. The `Bundler` option is an easier option for migration as it prevents a huge refactor of our code.

Let's first look at the difference between `NodeNext` and `Bundler`.

### Difference between `NodeNext` and `Bundler`

While the two options mostly behave the same in the ways we care, and both of them are recommended options for modern projects, yet they do have differences. Refer to this PR for more: https://github.com/microsoft/TypeScript/pull/51669.

The `NodeNext` option requires all relative imports in ESM to include a file extension, meaning all the following import examples should be refactored to contain a file extension `.js` at the end of the path:
```ts
import { a } from './a';      // Invalid
import { a } from './a.js';   // Correct

import { b } from 'pkg/b';    // Invalid
import { b } from 'pkg/b.js'; // Correct
```

The `Bundler` option, however, doesn't treat file extensions as mandatory and will leave the problem to our bundler: `tsup`. While bundling, `tsup` packs all source code into a single JS file, which means in our bundled code, there are no local imports, only the package imports remain:
```ts
import { a } from './a';      // This line doesn't exist in the bundled code
import { b } from 'pkg/b';    // This line exists in the bundled code
```

So with the `Bundler` option, we only need to deal with the package imports:
```ts
import { b } from 'pkg/b';    // Invalid
import { b } from 'pkg/b.js'; // Correct
```

### File extension are just promises

> For more discussions: https://github.com/microsoft/TypeScript/issues/49083

Another thing about the file extension being mandatory is that our bundlers (`tsc` and `tsup`) refuse to transpile the file extension of the imports from `.ts` to `.js` for us while bundling. That is, if we decide to add file extensions to the relative imports in our source code, even if we're importing a `.ts` file, we have to use `.js` like they're some sort of promises:
```ts
import { a } from './a.js';   // There's only an "a.ts" file in the directory
```

However, adding `.js` to package imports feels okay because they are usually just `.js` files:
```ts
import { b } from 'pkg/b.js'; // The "pkg/b.js" file actually exists
```

### The costs of migrating to ESM-first packages

What if we just want to migrate to specifying the `"type": "module"` and `"moduleResolution": "NodeNext"` together, as they are the most recommended settings for ESM-first packages?

Here's a branch that has done most of the refactor work, where all relative imports are refactored to include a file extension, and `"type": "module"` is added to the `package.json` to specify explicitly that the package is an ESM-first library:
- Commit: [[ckb-cell/rgbpp-sdk@fedb0f6]()](https://github.com/ckb-cell/rgbpp-sdk/commit/fedb0f6c4bad9056c330104087d9a357bfc7e9da)
- Snapshot: [`0.0.0-20240709190558`](https://www.npmjs.com/package/rgbpp/v/0.0.0-snap-20240709190558)
- Changes:
  1. Add file extension `.js` to all relative imports.
  2. Add `"type": "module"` to `package.json` to tag explicitly that this is an ESM-first package.
  3. Fix some type issues or weird incompatibilities found in the `rgbpp-sdk/ckb` lib to prevent bundling failures.
- Test results:
  - Tested in https://github.com/ShookLyngs/test-rgbpp-pure-esm in both ESM/CJS.
  - Tested in a CLI project to bundle it into a single executable application.

### Use `NodeNext` without migration

While testing everything, I find that we could specify the `NodeNext` option as the module resolution and NOT set `"type": "module"` in the `package.json`, which means by default our packages are just treated as CJS. With this combination, the TSC actually reports no error about our relative imports, even if they have no file extension included.

The problems are:
1. I'm not sure if this is a valid thing to do, or if it is simply an unresolved bug. Any documentation about this?
2. If we're using this combination, what's the difference from using the `Bundler` option? As they both don't give warnings from the TSC and IDE, we still need to validate the packages after bundling them.

### Conclusion

After everything, these are my thoughts about the combinations:

- Specify the packages as `"type": "module"`, and specify `"moduleResolution": "NodeNext"`
  - We have to refactor most of the code, and the file extensions in the imports are not what they look like.
  - If we decide to overcome the downsides, this is the recommended choice by the community.
- Keep the packages as `"type": "commonjs"`, and specify `"moduleResolution": "NodeNext"`
  - Behaves the same as `Bundler`. Since we're using `tsup`, I think this other option is more appropriate.
  - I'm not sure if this combination is a bug of TSC or if it is simply a valid choice.
- Keep the packages as `"type": "commonjs"`, and specify `"moduleResolution": "Bundler"`
  - The bundled code is not promised to work at runtime because file extensions for package imports are not mandatory, therefore lacking the process to validate the source code.
  - If we treat `tsup` as a bundler, we can use this option.

Reply:
Have you tried with the `"module": "node16"` and `"moduleResolution": "nodenext"` of the `tsconfig.json`.

See https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#im-writing-a-library

- module: "node16". When a codebase is compatible with Node.js’s module system, it almost always works in bundlers as well. If you’re using a third-party emitter to emit ESM outputs, ensure that you set "type": "module" in your package.json so TypeScript checks your code as ESM, which uses a stricter module resolution algorithm in Node.js than CommonJS does. As an example, let’s look at what would happen if a library were to compile with "moduleResolution": "bundler":

```
export * from "./utils";
```

Assuming ./utils.ts (or ./utils/index.ts) exists, a bundler would be fine with this code, so "moduleResolution": "bundler" doesn’t complain. Compiled with "module": "esnext", the output JavaScript for this export statement will look exactly the same as the input. If that JavaScript were published to npm, it would be usable by projects that use a bundler, but it would cause an error when run in Node.js:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '.../node_modules/dependency/utils' imported from .../node_modules/dependency/index.js
Did you mean to import ./utils.js?
```

Reply:
Yes, I've tried it, a TS error will be thrown if we don't use them in pairs (that's why I didn't mention it):

```
error TS5110: Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'
```

FYI, Node16 and NodeNext are currently identical, according to the documentation: https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext.